src/Microsoft.AspNetCore.Http/StreamPipeWriter.cs

```diff
@@ -0,0 +1,314 @@
+        }
+
+        private void Cancel()
+        {
+            InternalTokenSource.Cancel();
+        }
+
+        private ValueTask<FlushResult> FlushAsyncInternal(CancellationToken cancellationToken = default)
+        {
+            CancellationTokenRegistration reg = new CancellationTokenRegistration();
+            if (cancellationToken.CanBeCanceled)
+            {
+                reg = cancellationToken.Register(state => ((StreamPipeWriter)state).Cancel(), this);
+            }
+
+            return WriteAndFlushAsync(reg, cancellationToken);
+        }
+
+        private async ValueTask<FlushResult> WriteAndFlushAsync(CancellationTokenRegistration reg, CancellationToken token)
+        {
+            // Write all completed segments and whatever remains in the current segment
+            // and flush the result.
+            using (reg)
+            {
+                try
+                {
+                    if (_completedSegments != null && _completedSegments.Count > 0)
+                    {
+                        var count = _completedSegments.Count;
+                        for (var i = 0; i < count; i++)
+                        {
+                            var segment = _completedSegments[0];
+#if NETCOREAPP2_2
+                            await _writingStream.WriteAsync(segment.Buffer.Slice(0, segment.Length), InternalTokenSource.Token);
+#elif NETSTANDARD2_0
+                            MemoryMarshal.TryGetArray<byte>(segment.Buffer, out var arraySegment);
+                            await _writingStream.WriteAsync(arraySegment.Array, 0, segment.Length, InternalTokenSource.Token);
+#else
+#error Target frameworks need to be updated.
+#endif
+                            _bytesWritten -= segment.Length;
+                            segment.Return();
+                            _completedSegments.RemoveAt(0);
+                        }
+                    }
+
+                    if (!_currentSegment.IsEmpty)
+                    {
+#if NETCOREAPP2_2
+                        await _writingStream.WriteAsync(_currentSegment.Slice(0, _position), InternalTokenSource.Token);
+#elif NETSTANDARD2_0
+                        MemoryMarshal.TryGetArray<byte>(_currentSegment, out var arraySegment);
+                        await _writingStream.WriteAsync(arraySegment.Array, 0, _position, InternalTokenSource.Token);
+#else
+#error Target frameworks need to be updated.
+#endif
+                        _bytesWritten -= _position;
+                        _position = 0;
+                    }
+
+                    await _writingStream.FlushAsync(_internalTokenSource.Token);
+
+                    return new FlushResult(isCanceled: false, IsCompletedOrThrow());
+                }
+                catch (OperationCanceledException)
+                {
+                    // Remove the cancellation token such that the next time Flush is called
+                    // A new CTS is created.
+                    lock (lockObject)
```

Comment:
Reading is not using lock and writing does, suspicious.

Reply:
It's used when accessing the property: https://github.com/aspnet/HttpAbstractions/pull/1065/files#diff-dcfa2ec5d0439ce1abf9d81d7f9fc302R41